File: back/app/api/v1/announcements.py
"""공고 조회 API (5주차 1차 작업 순서 8)

- GET /announcements       : 검색어·상태·기관·출처 필터 + 정렬 + 페이지네이션
- GET /announcements/{id}  : 공고 상세 (없으면 404)

주의: `status`는 아직 원본 데이터(source마다 다른 문자열, 예: "Y"/"N", "일반공고" 등)를
그대로 저장하고 있어서(app/services/collector.py 참고), FE의 접수중/마감 같은 정규화된
값과는 다를 수 있습니다. 상태값을 통일하는 작업은 이번 범위에 포함하지 않았고,
지금은 저장된 값 그대로 필터링만 됩니다 (FE 연동 시 다시 논의 필요).
"""
import uuid
from datetime import date

from fastapi import APIRouter, Depends, Query
from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.core.errors import AppError
from app.db.models import Announcement
from app.db.session import get_db

router = APIRouter(tags=["announcements"])

# 5-1plan.md "정렬 권장 기준" 표 그대로 반영
SORT_OPTIONS = {
    "latest": (Announcement.reception_start.desc(), Announcement.id.desc()),
    "deadline": (Announcement.reception_end.asc().nulls_last(), Announcement.id.asc()),
    "title": (Announcement.title.asc(), Announcement.id.asc()),
}


def _dday(reception_end: date | None) -> int | None:
    """오늘 기준 마감일까지 남은 일수. 마감일 정보가 없으면 null."""
    if reception_end is None:
        return None
    return (reception_end - date.today()).days


def _serialize(row: Announcement) -> dict:
    return {
        "id": str(row.id),
        "source": row.source,
        "external_id": row.external_id,
        "title": row.title,
        "department": row.department,
        "reception_start": row.reception_start,
        "reception_end": row.reception_end,
        "status": row.status,
        "detail_url": row.detail_url,
        "summary": row.summary,
        "collected_at": row.collected_at,
        "dday": _dday(row.reception_end),
    }


@router.get("/announcements")
def list_announcements(
    q: str | None = Query(None, description="제목 검색어 (부분 일치)"),
    status: str | None = Query(None, description="공고 상태 필터 (저장된 값과 정확히 일치해야 함)"),
    department: str | None = Query(None, description="기관/부서 필터"),
    source: str | None = Query(None, description="수집 출처 필터 (kstartup, narajangteo, msit)"),
    sort: str = Query("latest", description="정렬 기준: latest | deadline | title"),
    page: int = Query(1, ge=1, description="페이지 번호 (1부터 시작)"),
    page_size: int = Query(20, ge=1, le=100, description="페이지당 개수"),
    db: Session = Depends(get_db),
):
    if sort not in SORT_OPTIONS:
        raise AppError(400, "INVALID_SORT", "sort는 latest, deadline, title 중 하나여야 합니다.")

    conditions = []
    if q:
        conditions.append(Announcement.title.ilike(f"%{q}%"))
    if status:
        conditions.append(Announcement.status == status)
    if department:
        conditions.append(Announcement.department == department)
    if source:
        conditions.append(Announcement.source == source)

    count_stmt = select(func.count()).select_from(Announcement)
    list_stmt = select(Announcement)
    for condition in conditions:
        count_stmt = count_stmt.where(condition)
        list_stmt = list_stmt.where(condition)

    total = db.execute(count_stmt).scalar_one()

    list_stmt = (
        list_stmt.order_by(*SORT_OPTIONS[sort]).offset((page - 1) * page_size).limit(page_size)
    )
    rows = db.execute(list_stmt).scalars().all()

    return {
        "success": True,
        "data": [_serialize(row) for row in rows],
        "meta": {"total": total, "page": page, "page_size": page_size},
    }


@router.get("/announcements/{announcement_id}")
def get_announcement(announcement_id: str, db: Session = Depends(get_db)):
    try:
        parsed_id = uuid.UUID(announcement_id)
    except ValueError:
        raise AppError(404, "ANNOUNCEMENT_NOT_FOUND", "존재하지 않는 공고입니다.")

    row = db.get(Announcement, parsed_id)
    if row is None:
        raise AppError(404, "ANNOUNCEMENT_NOT_FOUND", "존재하지 않는 공고입니다.")

    return {"success": True, "data": _serialize(row)}

File: front/src/lib/announcements.ts
import { api } from './api'
import type { Announcement, StatusType } from '@/types'

/** back/app/api/v1/announcements.py의 _serialize()가 실제로 내려주는 필드. */
interface ApiAnnouncement {
  id: string
  source: string
  external_id: string
  title: string
  department: string | null
  reception_start: string | null
  reception_end: string | null
  status: string | null
  statusLabel: StatusType | null
  detail_url: string | null
  summary: string | null
  collected_at: string
  dday: number | null
}

interface ListMeta {
  total: number
  page: number
  page_size: number
}

const SOURCE_LABELS: Record<string, string> = {
  kstartup: 'K-Startup',
  narajangteo: '나라장터',
  msit: '과기정통부',
}

/**
 * BE 응답을 FE `Announcement` 형태로 맞춘다. BE는 org/announcementType/field/postedDate/
 * receiptDate/deadlineTime/budget/contact/projectName/relatedKeywords 같은 필드를 아직
 * 내려주지 않으므로, 화면이 깨지지 않도록 있는 값(department/source/summary 등)으로 채우거나
 * 빈 값을 넣는다. (docs/fe/5th_wk_FE_연동작업.md 1절 참고)
 */
export function mapAnnouncement(raw: ApiAnnouncement): Announcement {
  const org = raw.department ?? '기관 정보 없음'
  const sourceLabel = SOURCE_LABELS[raw.source] ?? raw.source
  return {
    id: raw.id,
    title: raw.title,
    org,
    department: org,
    announcementType: sourceLabel,
    announceType: sourceLabel,
    field: '',
    status: raw.statusLabel ?? '접수중',
    postedDate: raw.reception_start ?? '-',
    receiptDate: raw.reception_start ?? '-',
    deadline: raw.reception_end ?? '기한 미정',
    deadlineTime: '-',
    dday: raw.dday,
    originalUrl: raw.detail_url ?? undefined,
    originalText: raw.summary ?? undefined,
    relatedKeywords: [],
  }
}

export interface AnnouncementQuery {
  q?: string
  /**
   * 키워드 매칭 모드 — 여러 개면 BE가 OR로 매칭한다(제목에 하나라도 포함되면 포함).
   * 대시보드 "매칭된 공고" 카운트·목록과 정확히 같은 집합을 만드는 용도.
   */
  keywords?: string[]
  statusLabel?: StatusType
  sort?: 'latest' | 'deadline' | 'title'
  page?: number
  pageSize?: number
}

export async function listAnnouncements(query: AnnouncementQuery) {
  const params = new URLSearchParams()
  if (query.q) params.set('q', query.q)
  for (const kw of query.keywords ?? []) params.append('keywords', kw)
  if (query.statusLabel) params.set('statusLabel', query.statusLabel)
  params.set('sort', query.sort ?? 'latest')
  params.set('page', String(query.page ?? 1))
  params.set('page_size', String(query.pageSize ?? 8))

  const { data, meta } = await api.get<ApiAnnouncement[]>(`/announcements?${params.toString()}`)
  return { items: data.map(mapAnnouncement), meta: meta as unknown as ListMeta }
}

export async function getAnnouncementDetail(id: string): Promise<Announcement> {
  const { data } = await api.get<ApiAnnouncement>(`/announcements/${id}`)
  return mapAnnouncement(data)
}
